google/cloud/bigquery/v2/minimal/internal/dataset_request.cc

```diff
@@ -0,0 +1,145 @@
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/bigquery/v2/minimal/internal/dataset_request.h"
+#include "google/cloud/common_options.h"
+#include "google/cloud/internal/absl_str_cat_quiet.h"
+#include "google/cloud/internal/debug_string.h"
+#include "google/cloud/internal/format_time_point.h"
+#include "google/cloud/internal/make_status.h"
+#include "google/cloud/status.h"
+#include "absl/strings/match.h"
+
+namespace google {
+namespace cloud {
+namespace bigquery_v2_minimal_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+namespace {
+
+std::string GetDatasetsEndpoint(Options const& opts) {
+  std::string endpoint = opts.get<EndpointOption>();
+
+  if (!absl::StartsWith(endpoint, "https://") &&
+      !absl::StartsWith(endpoint, "http://")) {
+    endpoint = absl::StrCat("https://", endpoint);
+  }
+  if (!absl::EndsWith(endpoint, "/")) absl::StrAppend(&endpoint, "/");
+  absl::StrAppend(&endpoint, "bigquery/v2/projects/");
+
+  return endpoint;
+}
+
+}  // namespace
+
+std::string GetDatasetRequest::DebugString(absl::string_view name,
+                                           TracingOptions const& options,
+                                           int indent) const {
+  return internal::DebugFormatter(name, options, indent)
+      .StringField("project_id", project_id_)
+      .StringField("dataset_id", dataset_id_)
+      .Build();
+}
+
+std::string ListDatasetsRequest::DebugString(absl::string_view name,
+                                             TracingOptions const& options,
+                                             int indent) const {
+  return internal::DebugFormatter(name, options, indent)
+      .StringField("project_id", project_id_)
+      .Field("all_datasets", all_datasets_)
+      .Field("max_results", max_results_)
+      .StringField("page_token", page_token_)
+      .StringField("filter", filter_)
+      .Build();
+}
+
+ListDatasetsRequest::ListDatasetsRequest(std::string project_id)
+    : project_id_(std::move(project_id)),
+      all_datasets_(false),
+      max_results_(0) {}
+
+StatusOr<rest_internal::RestRequest> BuildRestRequest(
+    GetDatasetRequest const& r) {
+  auto const& opts = internal::CurrentOptions();
+
+  rest_internal::RestRequest request;
+  if (r.project_id().empty()) {
+    return internal::InvalidArgumentError(
+        "Invalid GetDatasetRequest: Project Id is empty", GCP_ERROR_INFO());
+  }
+  if (r.dataset_id().empty()) {
+    return internal::InvalidArgumentError(
+        "Invalid GetDatasetRequest: Dataset Id is empty", GCP_ERROR_INFO());
+  }
```

Comment:
FWIW, I would vote against these sort of checks, unless they are necessary. I imagine the server would return the same sort of error. Less code is better.